modules/fedimint-wallet-client/src/pegin_monitor.rs

```diff
@@ -0,0 +1,488 @@
+                    // In tests, we basically mine all blocks right away
+                    Some(Duration::from_millis(1))
+                } else {
+                    Some(Duration::from_secs(60 * num_blocks_needed))
+                }
+            }
+            // Once anything has been claimed, there's no reason to claim again automatically,
+            // and it's undesirable due to privacy reasons.
+            // Users can possibly update the underlying record via other means to force a check on
+            // demand.
+            CheckOutcome::Claimed { .. } | CheckOutcome::AlreadyClaimed => None,
+        }
+    }
+
+    /// Desired retry delay for a bunch of outcomes.
+    ///
+    /// This time is intended to be persisted in the database.
+    ///
+    /// None means "no need to check anymore".
+    fn retry_delay_vec(outcomes: &[CheckOutcome], creation_time: SystemTime) -> Option<Duration> {
+        // If the address was allocated, but nothing was ever received or even detected
+        // on it yet, check again in time proportional to the age of the
+        // address.
+        if is_running_in_test_env() {
+            // When testing we usually send deposits right away, so check more aggressively.
+            return Some(Duration::from_millis(100));
+        } else if outcomes.is_empty() {
+            let now = time::now();
+            let age = now.duration_since(creation_time).unwrap_or_default();
+            return Some(age / 2);
+        }
+
+        // The delays is the minimum retry delay.
+        let mut min = None;
+
+        for outcome in outcomes {
+            min = match (min, outcome.retry_delay()) {
+                (None, time) => time,
+                (Some(min), None) => Some(min),
+                (Some(min), Some(time)) => Some(cmp::min(min, time)),
+            };
+        }
+
+        min
+    }
+
+    fn get_claimed_now_outpoints(outcomes: &[CheckOutcome]) -> Vec<bitcoin::OutPoint> {
+        let mut res = vec![];
+        for outcome in outcomes {
+            if let CheckOutcome::Claimed { outpoint } = outcome {
+                res.push(*outpoint);
+            }
+        }
+
+        res
+    }
+}
+
+/// Query via btc rpc for a history of an address derived with `tweak_idx` and
+/// claim any peg-ins that are ready.
+///
+/// Return a list of [`CheckOutcome`]s for each matching output.
+#[instrument(skip_all, fields(tweak_idx))]
+async fn check_idx_pegins(
+    data: &WalletClientModuleData,
+    tweak_idx: TweakIdx,
+    btc_rpc: &DynBitcoindRpc,
+    module_rpc: &DynModuleApi,
+    db: &Database,
+    client_ctx: &ClientContext<WalletClientModule>,
+) -> Result<Vec<CheckOutcome>, anyhow::Error> {
+    let current_consensus_block_count = module_rpc.fetch_consensus_block_count().await?;
+    let (script, address, tweak_key, operation_id) = data.derive_peg_in_script(tweak_idx);
+    btc_rpc.watch_script_history(&script).await?;
+
+    let history = btc_rpc.get_script_history(&script).await?;
```

Comment:
Not a blocker for this PR, but we may want to consider updating the esplora impl for paging since this will only return 25 items.

Reply:
Yeah, we probably want to read all pages inside the impl of that function (with some very high limit maybe, to prevent abuse). We never call these with high performance requirements, and I doubt we'll want to handle pagination in the higher level loop.